unionfs-1.3/man/unionctl.2
.\" Process with groff -man -Tascii unionfs.1
.\"
.TH UNIONFS 2 "May 2006" Linux
.SH NAME
Unionfs \- a unification file system for Linux
.SH SYNOPSIS

.nf
.B #include <sys/ioctl.h>
.B #include <unionfs.h>
.sp
.BI "int ioctl(int " fd ", int " action ", void *" args );
.fi
.SH DESCRIPTION
Unionfs uses 
.B ioctl
for branch management. The three arguments passed are the file descriptor,
.IR fd ,
which unless otherwise noted, is for the root of the union;
.I action
which specifies what branch management operation is desired; and
.I args
which is a pointer to action-dependent arguments.

The following actions are defined:
.TP
.B UNIONFS_IOCTL_ADDBRANCH
When this action is specified, the directory by the
.B unionfs_addbranch_args
structure will be added to the union.

.sp
.nf
struct unionfs_addbranch_args {
.in +8
unsigned int ab_branch;
char *ab_path;
unsigned int ab_perms;
.in -8
};
.fi

.I ab_branch
is the requested branch number. All the branches already in the union with
branch number greater than or equal to
.I ab_branch
will get their branch numbers incremented by one, and the new branch will fill
the empty position.

.I ab_path
is the path for the new branch. There is no default, and it must be specified.

.I ab_perms
are the permissions for the new branch. The default is
.BR "MAY_READ | MAY_WRITE" .

.TP
.B UNIONFS_IOCTL_DELBRANCH
Branch removal is currently done in a slightly different way. One needs to first
allocate space for 3 arguments passed to mount(2).

.sp
.nf
unsigned long remount_args[3];
.fi

One allocated, they need to be filled in appropriately. For example, if we
want to remove branch number
.B bnum
from the union mounted on
.BR mount_path ,
we set the arguments to:

.sp
.nf
remount_args[0] = UNIONFS_REMOUNT_MAGIC;
remount_args[1] = UNIONFS_IOCTL_DELBRANCH;
remount_args[2] = bnum;
.fi

One filled in, we can pass it to mount(2). For detailed explanation of the
arguments, see the mount manual page.

.sp
.nf
mount("unionfs", mount_path, "unionfs", MS_REMOUNT | MS_MGC_VAL, remount_args);
.fi

.TP
.B UNIONFS_IOCTL_RDWRBRANCH
This action allows the permissions on branches of an already mounted union to
be changed. The
.I args
argument is a pointer to a
.B unionfs_rdwrbranch_args
structure which contains the necessary information.

.sp
.nf
struct unionfs_rdwrbranch_args {
.in +8
unsigned int rwb_branch;
unsigned int rwb_perms;
.in -8
};
.fi

.I rwb_branch
is number of the branch that the permission change will take effect.

.I rwb_perms
is the permission mask. It consists of
.B MAY_READ
,
.B MAY_WRITE
, and
.B MAY_NFSRO.
The
.B MAY_NFSRO
bit forces Unionfs to ignore -EACCES on NFS mounted file systems.

.TP
.B UNIONFS_IOCTL_QUERYFILE
This action allows the caller to find out which branch a particular file is
taken from. The file descriptor,
.BR fd ,
is obtained by open(2)ing the file in question through unionfs, along with a
pointer to an
.B fd_set
as 
.IR args .

One can then use
.BR FD_ISSET ,
see select(2), to check which "descriptor" is set in
.IR args .
The index passed to
.B FD_ISSET
is equal to the branch number that contains the file identified by
.IR fd .

.SH AUTHORS
Josef Sipek <jsipek@cs.sunysb.edu>,
Erez Zadok <ezk@cs.sunysb.edu>
.SH "SEE ALSO"
.BR unionctl(8),
.BR unionfs(4),
.BR uniondbg(8),
http://unionfs.filesystems.org/

\"  LocalWords:  groff Tascii unionfs unionctl rw ro Erez uniondbg


unionfs-1.3/NEWS
* Unionfs 1.3 (intended for use with 2.6.17)
- Fixed makefile release & build target
- Fixed up uniondbg & unionimap man pages
- Updated RPM specfile
- Implemented export functions (for NFS)
- Updated to compile with 2.6.17

* Unionfs 1.2 (intended for use with 2.6.16)
- Added basic mmap support (experimental, off by default)
- Added check to see if its necessary to run the delete_all
  regression tests
- Cleaned up list/get xattr code
- Created documentation directory
- Fixed unionfs_permission on reiserfs and XFS
- Fixed link-unlink issue (i_nlink going down to zero)
- Fixed Bug #501: mv dir ..
- On error, NULL the pointers after free - Fixes nasty mount bug
- Moved userspace utilities into utils subdirectory
- Updated to work with the new mutex subsystem
- Simplified and cleaned up debugging support (use -DUNIONFS_DEBUG to
  _enable_ debugging)
- Header files simplified

* Unionfs 1.1.5
- Backport: Corrected sample mount commands
- Backport: Fixed link-unlink issue
- Backport: Fixed refcounting leak (if a branch doesn't exist during mount,
  the previously looked up branches were not freed properly)
- Backport: Fixed inode_permission returning EROFS when it wasn't supposed
  to
- Backport: Fixed bug #501 (mv dir ..)

* Unionfs 1.1.4
- Backport: On error, NULL the pointers after free - Fixes nasty mount bug
- Backport: Fixed unionfs_permission on reiserfs and XFS
- Backport: Added check to see if its necessary to run the delete_all
  regression tests
- Added explicit check for kernel versions newer than 2.6.15
(If you want to use 2.6.16 or newer, look at Unionfs 1.2.x)

* Unionfs 1.1.3
- Simplified delete=whiteout
- Switched default delete mode to the simpler whiteout mode
- Disable delete=all by default (enable with -DUNIONFS_DELETE_ALL)
- Removed all copy up modes except preserve (the old default)
- Unlink partially copied up files
- Compartmentalize malloc debugging
- Fix uid/gid/mode not being reset when a whiteout is recreated
- Fix stale atime/mtime bugs for recreated whiteouts
- Cleaned up peristent inode code a bit (still unstable)

* Unionfs 1.1.2
- Added inode refcounting debugging tool
- Fixed race between lookup and d_free
- Fixed double free/unchecked malloc
- Fixed permission bug, creat/open truncates the running executable
- Fixed many reference counting bugs
- Moved code from fist.h into unionfs*.h
- Removed unneeded UNIONFS_XATTR
- Removed ASSERTs in favor of BUG_ONs
- Removed FISTBUG
- Rename dentry locking changed to use 2.6 calls
- Replaced NFS_SECURITY_HOLE by nfsro branch option
- Cleaned up code to "pass" a Sparse run
- RPM spec file updated
- Miscellaneous cleanups

* Unionfs 1.1.1
- Directory reading is fixed
- Improved locking for branch manipulation operations
- Fix for removing an opaque directory
- Some cleanups and error handling fixes to the copyup code

* Unionfs 1.1.0
- Removed excess 2.4 code (Unionfs 1.1.0 is for 2.6 only)
- Removed file locking code, as the VFS does a better job of it.
- Fixed off by one errors when kmallocing names in inode.c.
- Fixed several deadlock bugs.
- Fixed several rename and rmdir bugs.
- Permissions checked on files that are on read-only file systems
- AMD64 fix for compiling on gentoo.
- Properly update generation when we combine copyup and reopening.
- Cleanedup Makefile to remove 2.4 clutter.
- Added documentation for the MODDIR option in the INSTALL file.

* Unionfs 1.0.14
Features:
- Dropped unmaintained setattr,diropaque,delete=first and mount flags
- Updated unionfs.4 to reflect default mount modes
- You can use the root of the Union to remove branches with unionctl
- Use official Debian packaging files.
- Linux 2.6.13 support.

Bug fixes:
- Fixed several dentry refcount bugs introduced by new deletion framework.
- Fixed uninitialized fd_set in the query ioctl.
- Branch reference counting now works across insertion and removal of branches.
  This prevents the branch counts from getting "confused".
- Handle "/" as a branch in unionctl.
- Removed static buffer for debug prints in favor of vprintk.
- NFS silly renames avoided during several cases.
- Attempting a write lock causes a copyup, so that the underying flock will
  work.
- Cleaned up mount option parsing.
- Improved link counting for directories.

* Unionfs 1.0.13
- Unionfs now handles deleting an open file (and several other deletion fixes).
  Thanks to the Knoppix people and Tomas Matejicek for finding these bugs.
- The NODEBUG flag has been renamed to UNIONFS_NDEBUG.
- Instead of allocating all of the lower-level pointers in dentries, inodes
  and files, we have a fixed number of them preallocated and malloc only if
  we exceed that number.
- Fixed a dentry reference count bug with UNIONFS_NDEBUG.
- Support for persistent inode numbers across mounts.
- Added a "which-branch" ioctl and corresponding --query to unionctl, so that
  it is possible to find out where a file or directory is coming from.
- In 2.6.11 and onwards, don't take the BKL for ioctls.

* Unionfs 1.0.12
- Copy-up is delayed until a write operation occurs, so spurious copy-ups are
  avoided.
- Improved copyup UID spoofing for mkdir
- Instructions and shell script for compiling Unionfs into a monolothic kernel.
- Added support for opaque directories, which mask the contents of
  lower-priority directories.  This makes recreating a removed directory an
  O(1)  operation instead of an O(n) operation, and also makes subsequent
  operations faster.  There were also several whiteout related fixes bundled
  in here.
- Fixed bug with whiteouts on Reiserfs
- Fixed bug when a file transitioned from a negative dentry to a positive
  dentry (i.e., if you add a branch that inserts a file you previously
  looked up, but did not exist)
- Fixed return code for extended attribute functions
- amd64 compile fixes
- Extended attribute copy-up
- Unionfs returns EXDEV instead of copying up a directory on move, this
  fixes a bug where Unionfs would "forget" the contents of a copied up
  directory after unmount.

unionfs-1.3/utils/unionctl.c
/*
 * Copyright (c) 2003-2006 Erez Zadok
 * Copyright (c) 2003-2006 Charles P. Wright
 * Copyright (c) 2005-2006 Josef Sipek
 * Copyright (c) 2005      Arun M. Krishnakumar
 * Copyright (c) 2005-2006 David P. Quigley
 * Copyright (c) 2003-2004 Mohammad Nayyer Zubair
 * Copyright (c) 2003      Puja Gupta
 * Copyright (c) 2003      Harikesavan Krishnan
 * Copyright (c) 2003-2006 Stony Brook University
 * Copyright (c) 2003-2006 The Research Foundation of State University of New York
 *
 * For specific licensing information, see the COPYING file distributed with
 * this package.
 *
 * This Copyright notice must be kept intact and distributed with all sources.
 */
/*
 *  $Id: unionctl.c,v 1.1 2006/05/31 02:51:04 dquigley Exp $
 */

#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/mount.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <limits.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <sys/types.h>

#include "unionfs.h"

#define MAY_READ 4
#define MAY_WRITE 2

extern int find_union(const char *path, char **options, char **actual_path,
		      int uniononly);
static char **branches;
static int *branchperms;
static const char *progname;

void __attribute__ ((__noreturn__)) __usage(int line);

#define usage() __usage(__LINE__)

void __attribute__ ((__noreturn__)) __usage(int line)
{
#ifdef DEBUG
	fprintf(stderr, "Line: %d\n", line);
#endif
	fprintf(stderr,
		"unionctl version: $Id: unionctl.c,v 1.1 2006/05/31 02:51:04 dquigley Exp $\n");
	fprintf(stderr, "Distributed with Unionfs " UNIONFS_VERSION "\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "unionctl UNION ACTION [arguments]\n");
	fprintf(stderr,
		"ACTION can be --add, --remove, --mode, --list, or --query.\nFurther arguments depend on ACTION.\n");
	fprintf(stderr,
		"\tunionctl UNION --add [ --before BRANCH | --after BRANCH ] [ --mode (rw|ro|nfsro) ] DIRECTORY\n");
	fprintf(stderr, "\tunionctl UNION --remove BRANCH\n");
	fprintf(stderr, "\tunionctl UNION --mode BRANCH (rw|ro|nfsro)\n");
	fprintf(stderr, "\tunionctl UNION --list\n");
	fprintf(stderr, "\tunionctl FILENAME --query\n");
	fprintf(stderr,
		"The unionctl man page has a more detailed description of its usage.\n");
	exit(EXIT_FAILURE);
}

static inline int parse_rw(char *p)
{
	if (strcmp(p, "ro") == 0)
		return MAY_READ;
	else if (strcmp(p, "nfsro") == 0)
		return MAY_READ | MAY_NFSRO;
	else if (strcmp(p, "rw") == 0)
		return MAY_READ | MAY_WRITE;
	else
		return 0;
}

static char **parse_options(char *options)
{
	char **ret = NULL;
	int i = 0;

	char *p;
	char *q;
	char *r;
	char *s, *t, *u;

	p = options;
	do {
		q = strchr(p, ',');
		if (q) {
			*q++ = '\0';
		}
		if (!strncmp(p, "dirs=", strlen("dirs="))) {
			r = p + strlen("dirs=");
			do {
				s = strchr(r, ':');
				if (s) {
					*s++ = '\0';
				}

				i++;
				ret = realloc(ret, sizeof(char *) * (i + 1));
				if (!ret) {
					perror("realloc()");
					exit(EXIT_FAILURE);
				}
				branchperms =
				    realloc(branchperms, sizeof(int) * i);
				if (!branchperms) {
					perror("realloc()");
					exit(EXIT_FAILURE);
				}

				t = strchr(r, '=');
				u = t + 1;
				if (!t || !u || !*u)
					goto err;
				*t = 0;
				branchperms[i - 1] = parse_rw(u);
				if (!branchperms[i - 1]) {
				      err:
					fprintf(stderr, "cannot parse '%s'\n",
						r);
					exit(EXIT_FAILURE);
				}
				ret[i - 1] = strdup(r);
				ret[i] = NULL;

				r = s;
			}
			while (r);
		}
		p = q;
	}
	while (p);

	branches = ret;
	return ret;
}

static int get_branch(char *path)
{
	char *end;
	int ret;

	ret = strtol(path, &end, 0);
	if (!*end) {
		return ret;
	}

	ret = strlen(path);
	if ((ret > 1) && (path[ret - 1] == '/')) {
		path[ret - 1] = '\0';
	}

	ret = 0;
	while (branches[ret]) {
		if (!strcmp(branches[ret], path)) {
			return ret;
		}
		ret++;
	}

	return -1;
}

static void dump_branches(const char *prefix)
{
	int i = 0;

	while (branches[i]) {
		char r, w, n;
		r = (branchperms[i] & MAY_READ) ? 'r' : '-';
		w = (branchperms[i] & MAY_WRITE) ? 'w' : '-';
		n = (branchperms[i] & MAY_NFSRO) ? 'n' : '-';
		printf("%s%s (%c%c%c)\n", prefix, branches[i], r, w, n);
		i++;
	}
}

#define ADD 1
#define REMOVE 2
#define MODE 3
#define LIST 4
#define	QUERY 5

int main(int argc, char *argv[])
{
	struct unionfs_addbranch_args addargs;
	struct unionfs_rdwrbranch_args rdwrargs;
	unsigned long remountdata[3];
	fd_set branchlist;
	struct stat st;

	int fd = -1;
	int ret, i;

	char *path, resolv_path[PATH_MAX], resolv_bp[PATH_MAX];
	char *options = NULL, *actual_path = NULL;
	int action;

	char *branchpath;
	int branchnum;
	int unionpos = 1;
	int modepos = 2;

	progname = argv[0];

	/* check that minimum number of args were specified */
	if (argc < 3)
		usage();

	if (argv[1][0] == '-' && argv[1][1] == '-') {
		modepos = 1;
		unionpos = 2;
	} else {
		modepos = 2;
		unionpos = 1;
	}

	if (realpath(argv[unionpos], resolv_path) == NULL) {
		perror("realpath()");
		exit(EXIT_FAILURE);
	}
	path = resolv_path;
	if (strcmp(path, "/") && (path[strlen(path) - 1] == '/')) {
		path[strlen(path) - 1] = '\0';
	}

	if (!strcmp(argv[modepos], "--add")) {
		action = ADD;
	} else if (!strcmp(argv[modepos], "--remove")) {
		action = REMOVE;
	} else if (!strcmp(argv[modepos], "--mode")) {
		action = MODE;
	} else if (!strcmp(argv[modepos], "--list")) {
		action = LIST;
	} else if (!strcmp(argv[modepos], "--query")) {
		action = QUERY;
	} else {
		usage();
	}

	if (stat(path, &st) == -1) {
		fprintf(stderr, "stat(%s): %s\n", path, strerror(errno));
		exit(EXIT_FAILURE);
	}

	if (find_union(path, &options, &actual_path, 1) < 0) {
		fprintf(stderr, "%s is not a valid union.\n", path);
		exit(EXIT_FAILURE);
	}
	branches = parse_options(options);
	if (!branches) {
		fprintf(stderr, "Could not parse options from /proc/mounts!\n");
		exit(EXIT_FAILURE);
	}

	/* open file on which ioctl will operate (that is actually any file in the union) */
	if (action != REMOVE) {
		fd = open(path, O_RDONLY);
		if (fd < 0) {
			fprintf(stderr, "open(%s): %s\n", path,
				strerror(errno));
			exit(EXIT_FAILURE);
		}
	}

	/* Parse the action's options into something usable, and do it. */
	switch (action) {
	case ADD:
		if (argc < 4)
			usage();

		/* Default values if the user leaves them unspecified. */
		branchnum = -1;
		addargs.ab_perms = MAY_READ | MAY_WRITE;
		branchpath = NULL;
		for (i = 3; i < argc; i++) {
			if (argv[i][0] == '-' && argv[i][1] == '-') {
				if (!strcmp(argv[i], "--before")) {
					i++;
					if (i == argc) {
						fprintf(stderr,
							"%s requires an argument!\n",
							argv[i - 1]);
						usage();
					}

					branchnum = get_branch(argv[i]);
					if (branchnum == -1) {
						fprintf(stderr,
							"%s is not a valid branch.\nThe current branch configuration is:\n",
							argv[i]);
						dump_branches("\t");
					}
				} else if (!strcmp(argv[i], "--after")) {
					i++;
					if (i == argc) {
						fprintf(stderr,
							"%s requires an argument!\n",
							argv[i - 1]);
						usage();
					}

					branchnum = get_branch(argv[i]);
					if (branchnum == -1) {
						fprintf(stderr,
							"%s is not a valid branch.\nThe current branch configuration is:\n",
							argv[i]);
						dump_branches("\t");
					}
					branchnum++;
				} else if (!strcmp(argv[i], "--mode")) {
					i++;
					if (i == argc) {
						fprintf(stderr,
							"%s requires an argument!\n",
							argv[i - 1]);
						usage();
					}

					addargs.ab_perms = parse_rw(argv[i]);
					if (!addargs.ab_perms) {
						fprintf(stderr,
							"Valid modes are ro, nfsro and rw you specified: \"%s\"\n",
							argv[i]);
						usage();
					}
				} else {
					fprintf(stderr, "Unknown option: %s\n",
						argv[i]);
					usage();
				}
			} else {
				int branchchk = -1;
				/* The options must go before the path */
				if ((i + 1) != argc) {
					fprintf(stderr,
						"The path of the branch to add must go after all options.\n");
					usage();
				}
				branchchk = get_branch(argv[i]);
				if (branchchk != -1) {
					fprintf(stderr,
						"%s already exists in the Union.\n",
						argv[i]);
					usage();
				}
				if (branchnum == -1)
					branchnum = 0;
				branchpath = argv[i];
			}
		}
		if (!branchpath) {
			fprintf(stderr,
				"You must specify the path to add into the union!\n");
			usage();
		}

		if (realpath(branchpath, resolv_bp) == NULL) {
			perror("realpath()");
			exit(EXIT_FAILURE);
		}

		addargs.ab_branch = branchnum;
		addargs.ab_path = resolv_bp;

		errno = 0;
		ret = ioctl(fd, UNIONFS_IOCTL_ADDBRANCH, &addargs);
		if (ret < 0) {
			switch (errno) {
			case E2BIG:
				fprintf(stderr,
					"Unionfs supports only %d branches.\n",
					FD_SETSIZE);
				break;
			}
			fprintf(stderr, "Failed to add %s into %s: %s",
				branchpath, path, strerror(errno));
			exit(EXIT_FAILURE);
		}
		break;
	case MODE:
		if (argc != 5) {
			usage();
		}

		branchnum = 3;
		rdwrargs.rwb_perms = parse_rw(argv[4]);
		if (!rdwrargs.rwb_perms) {
			branchnum = 4;
			rdwrargs.rwb_perms = parse_rw(argv[3]);
			if (!rdwrargs.rwb_perms) {
				usage();
				exit(EXIT_FAILURE);
			}
		}

		if (realpath(argv[branchnum], resolv_bp) == NULL) {
			perror("realpath()");
			exit(EXIT_FAILURE);
		}
		branchpath = resolv_bp;

		/* Set a branches writeable status. */
		rdwrargs.rwb_branch = get_branch(branchpath);
		if (rdwrargs.rwb_branch == -1) {
			fprintf(stderr,
				"%s is not a valid branch.\nThe current branch configuration is:\n",
				branchpath);
			dump_branches("\t");
			exit(EXIT_FAILURE);
		}

		ret = ioctl(fd, UNIONFS_IOCTL_RDWRBRANCH, &rdwrargs);
		if (ret < 0) {
			fprintf(stderr,
				"Failed to set permissions for %s in %s: %s",
				branchpath, path, strerror(errno));
			exit(EXIT_FAILURE);
		}

		goto out;
		break;
	case REMOVE:
		if (argc != 4)
			usage();

		if (realpath(argv[3], resolv_bp) == NULL) {
			perror("realpath()");
			exit(EXIT_FAILURE);
		}
		branchpath = resolv_bp;

		branchnum = get_branch(branchpath);
		if (branchnum == -1) {
			fprintf(stderr,
				"%s is not a valid branch.\nThe current branch configuration is:\n",
				branchpath);
			dump_branches("\t");
			exit(EXIT_FAILURE);
		}

		errno = 0;
		remountdata[0] = UNIONFS_REMOUNT_MAGIC;
		remountdata[1] = UNIONFS_IOCTL_DELBRANCH;
		remountdata[2] = branchnum;
		ret =
		    mount("unionfs", actual_path, "unionfs",
			  MS_REMOUNT | MS_MGC_VAL, remountdata);
		if (ret < 0) {
			fprintf(stderr, "Failed to remove %s from %s: %s",
				branchpath, path, strerror(errno));
			exit(EXIT_FAILURE);
		}
		break;
	case LIST:
		dump_branches("\t");
		break;

	case QUERY:
		if (argc != 3) {
			usage();
		}

		if ((fd = open(argv[unionpos], O_RDONLY)) < 0) {
			fprintf(stderr,
				"Unable to open file %s : %s",
				argv[unionpos], strerror(errno));
			exit(EXIT_FAILURE);
		}

		ret = ioctl(fd, UNIONFS_IOCTL_QUERYFILE, &branchlist);
		if (ret < 0) {
			fprintf(stderr,
				"Unable to retrieve list of branches for file %s : %s\n",
				argv[unionpos], strerror(errno));
			exit(EXIT_FAILURE);
		}

		for (i = 0; i <= ret; i++) {
			char r, w, n;
			r = (branchperms[i] & MAY_READ) ? 'r' : '-';
			w = (branchperms[i] & MAY_WRITE) ? 'w' : '-';
			n = (branchperms[i] & MAY_NFSRO) ? 'n' : '-';

			if (FD_ISSET(i, &branchlist))
				printf("%s\t%s (%c%c%c)\n", argv[unionpos],
				       branches[i], r, w, n);
		}
		break;
	}

      out:
	if (fd != -1)
		close(fd);
	exit(EXIT_SUCCESS);
}

/*
 *
 * vim:shiftwidth=8
 * vim:tabstop=8
 *
 * For Emacs:
 * Local variables:
 * c-basic-offset: 8
 * c-comment-only-line-offset: 0
 * c-offsets-alist: ((statement-block-intro . +) (knr-argdecl-intro . 0)
 *              (substatement-open . 0) (label . 0) (statement-cont . +))
 * indent-tabs-mode: t
 * tab-width: 8
 * End:
 */
